cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/services/AppFabricServer.java

```diff
@@ -190,5 +194,6 @@ protected void shutDown() throws Exception {
     programRuntimeService.stopAndWait();
     schedulerService.stopAndWait();
     adapterService.stopAndWait();
+    notificationService.stopAndWait();
```

Comment:
all these may throw exceptions. In that case, the following ones won't be stopped. Not good. I know it's not your code.